src/aiida/common/utils.py

```diff
@@ -608,3 +608,27 @@ def format_directory_size(size_in_bytes: int) -> str:
 
     # Format the size to two decimal places
     return f'{converted_size:.2f} {prefixes[index]}'
+
+
+def batch_iter(
+    iterable: Iterable[Any], size: int, transform: Optional[Callable[[Any], Any]] = None
+) -> Iterable[Tuple[int, List[Any]]]:
```

Comment:
Let's add a bit more type safety by ensuring that the passed callback accepts the type produced by Iterable :nerd_face: 

```suggestion

T = TypeVar('T')

def batch_iter(
    iterable: Iterable[T], size: int, transform: Callable[[T], Any] | None = None
) -> Iterable[tuple[int,list[Any]]]:
```

We could then (I think) provide the following overloads to get rid of `Any` in the return type

```
@overload
def batch_iter(iterable: Iterable[T], size: int, transform: None = None) -> Iterable[tuple[int,list[T]]]: ...

R = TypeVar('R')
@overload
def batch_iter(iterable: Iterable[T], size: int, transform: Callable[[T], R]) -> Iterable[tuple[int,list[R]]]: ...
```

Reply:
> Let's add a bit more type safety by ensuring that the passed callback accepts the type produced by Iterable 🤓

This is Python, why would we care about type safety? /s

Reply:
Done, thanks for the suggestion! Again, learning a lot :)